python/feldera/rest/_httprequests.py

```diff
@@ -1,231 +1,369 @@
+        :param body: The HTTP request body.
+        :param content_type: The value for `Content-Type` HTTP header. "application/json" by default.
+        :param params: The query parameters part of this request.
+        :param stream: True if the response is expected to be a HTTP stream.
+        :param serialize: True if the body needs to be serialized to JSON.
+        :param max_retries: Maximum number of retry attempts.
+        """
+        self.headers["Content-Type"] = content_type
+
+        prev_resp: requests.Response | None = None
+
+        try:
+            conn_timeout = self.config.connection_timeout
+            timeout = self.config.timeout
+            headers = self.headers
+
+            request_path = self.config.url + "/" + self.config.version + path
+
+            logging.debug(
+                "sending %s request to: %s with headers: %s, and params: %s",
+                http_method.__name__,
+                request_path,
+                str(headers),
+                str(params),
+            )
+
+            for attempt in range(max_retries):
+                if http_method.__name__ == "get":
+                    request = http_method(
+                        request_path,
+                        timeout=(conn_timeout, timeout),
+                        headers=headers,
+                        params=params,
+                        stream=stream,
+                        verify=self.requests_verify,
+                    )
+                elif isinstance(body, bytes):
+                    request = http_method(
+                        request_path,
+                        timeout=(conn_timeout, timeout),
+                        headers=headers,
+                        data=body,
+                        params=params,
+                        stream=stream,
+                        verify=self.requests_verify,
+                    )
+                else:
+                    request = http_method(
+                        request_path,
+                        timeout=(conn_timeout, timeout),
+                        headers=headers,
+                        data=json_serialize(body) if serialize else body,
+                        params=params,
+                        stream=stream,
+                        verify=self.requests_verify,
+                    )
+
+                prev_resp = request
+
+                try:
+                    resp = self.__validate(request, stream=stream)
+                    logging.debug("got response: %s", str(resp))
+                    return resp
+                except FelderaAPIError as err:
+                    # Handle 502 with intelligent health monitoring
+                    if err.status_code == 502:
+                        if self._handle_502_with_health_check(
+                            path, attempt, max_retries
+                        ):
+                            continue
+                        else:
+                            raise
```

Comment:
do we even get change to use 2nd attempt?

say we are at attempt 1, and health check returned False, we would raise the error without even retrying!

also maybe last attempt was for other error, do we still want to use same _attempt_ variable?

Reply:
>as we determine backoff interval in health check

this is for the case of spurious 502s that aren't realted to upgrades in this case it might attempt it again